### PR TITLE
[ADD] account_cost_center 8.0

### PR DESCRIPTION
**Pull request by astirpe**
Adding module Cost Center.
This module allows the user to link every invoice line to a cost center providing an extra dimension for the analysis.

### CODE REVIEW

**Review by nielsmiddeldorp — APPROVED** [review]

**Comment by andreschenkels** [comment]
:+1:  Works for me

File: account_cost_center/__openerp__.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
+# -*- coding: utf-8 -*-
+# Copyright 2015 Onestein (<http://www.onestein.eu>)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+{
+    'name': 'Costcenter',
+    'images': ['static/description/main_screenshot.png'],
+    'summary': """Costcenter information for invoice lines""",
```
Comment (hbrunn) [inline account_cost_center/__openerp__.py]: don't use multiline strings when not necessary

File: account_cost_center/models/account_cost_center.py [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from openerp import models, fields
+
+
+class AccountCostCenter(models.Model):
+    _name = 'account.cost.center'
+    _description = 'Account Cost Center'
+
+    name = fields.Char(string='Title', required=True, size=64)
```
Comment (hbrunn) [inline account_cost_center/models/account_cost_center.py]: `size` is obsolete for char fields, better remove them

File: account_cost_center/models/account_invoice.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+        help="Default Cost Center")
+
+    @api.model
+    def line_get_convert(self, line, part, date):
+        res = super(AccountInvoice, self).line_get_convert(line, part, date)
+        if line.get('cost_center_id'):
+            res['cost_center_id'] = line['cost_center_id']
+        return res
+
+    def fields_view_get(self, cr, uid, view_id=None, view_type='form',
```
Comment (hbrunn) [inline account_cost_center/models/account_invoice.py]: why no new api here?

File: account_cost_center/models/account_invoice.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+                        context=None, toolbar=False, submenu=False):
+        res = super(AccountInvoice, self).fields_view_get(
+            cr, uid, view_id=view_id, view_type=view_type,
+            context=context, toolbar=toolbar, submenu=submenu)
+        if not context:
+            context = {}
+        if not context.get('cost_center_default', False):
+            if view_type == 'form':
+                view_obj = etree.XML(res['arch'])
+                invoice_line = view_obj.xpath("//field[@name='invoice_line']")
```
Comment (hbrunn) [inline account_cost_center/models/account_invoice.py]: you can replace this hack with https://github.com/OCA/server-tools/tree/8.0/base_view_inheritance_extension by now

File: account_cost_center/models/account_invoice_line.py [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+from openerp import models, fields, api
+
+
+class AccountInvoiceLine(models.Model):
+    _inherit = 'account.invoice.line'
+
+    @api.model
+    def _default_cost_center(self):
+        return self.env['account.cost.center'].browse(
+            self._context.get('cost_center_id', None))
```
Comment (hbrunn) [inline account_cost_center/models/account_invoice_line.py]: `None` already is the default for `dict.get`

File: account_cost_center/security/ir.model.access.csv [hunk 6]
```diff
@@ -0,0 +1,4 @@
+id,name,model_id:id,group_id:id,perm_read,perm_write,perm_create,perm_unlink
+access_account_cost_center_invoice,account.cost.center,model_account_cost_center,account.group_account_invoice,1,1,1,1
+access_account_cost_center_user,account.cost.center,model_account_cost_center,account.group_account_user,1,1,1,1
```
Comment (hbrunn) [inline account_cost_center/security/ir.model.access.csv:3]: is it really a good idea to give the less privileged users full permissions on cost centers?

File: account_cost_center/views/account_cost_center.xml [hunk 7]
```diff
@@ -0,0 +1,79 @@
+<?xml version="1.0"?>
+<openerp>
+    <data>
+
+        <!-- Cost Center views -->
+        <record id="account_cost_center_form" model="ir.ui.view">
+            <field name="name">account.cost.center.form</field>
```
Comment (hbrunn) [inline account_cost_center/views/account_cost_center.xml]: let the orm generate a name if you don't pass anything informative

File: account_cost_center/views/account_cost_center.xml [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+            <field name="model">account.cost.center</field>
+            <field name="arch" type="xml">
+                <form string="Cost Center" version="7.0">
+                    <sheet>
+                        <group>
+                            <div class="oe_title">
+                                <div class="oe_edit_only">
+                                    <label for="name" string="Title"/>
```
Comment (hbrunn) [inline account_cost_center/views/account_cost_center.xml]: it suffices to attach the class to the label

File: account_cost_center/views/account_cost_center.xml [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+                        </group>
+                        <group name="main_group">
+                            <group name="main_group1">
+                                <field name="code"/>
+                            </group>
+                            <group name="main_group2">
+                            </group>
+                        </group>
+                        <notebook name="notebook">
+                        </notebook>
```
Comment (hbrunn) [inline account_cost_center/views/account_cost_center.xml:27]: empty elements on reserve?

File: account_cost_center/views/account_cost_center.xml [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+                    </sheet>
+                </form>
+            </field>
+        </record>
+
+        <record id="account_cost_center_tree" model="ir.ui.view">
+            <field name="name">account.cost.center.tree</field>
```
Comment (hbrunn) [inline account_cost_center/views/account_cost_center.xml]: see above

File: account_cost_center/views/account_cost_center.xml [hunk 11]
```diff
@@ -0,0 +1,79 @@
+<?xml version="1.0"?>
+<openerp>
+    <data>
+
+        <!-- Cost Center views -->
+        <record id="account_cost_center_form" model="ir.ui.view">
+            <field name="name">account.cost.center.form</field>
+            <field name="model">account.cost.center</field>
+            <field name="arch" type="xml">
+                <form string="Cost Center" version="7.0">
```
Comment (hbrunn) [inline account_cost_center/views/account_cost_center.xml]: the `string` and `version` attributes don't do anything here imho

File: account_cost_center/views/account_cost_center.xml [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+            </field>
+        </record>
+
+        <record id="account_cost_center_action" model="ir.actions.act_window">
+            <field name="name">Cost Centers</field>
+            <field name="type">ir.actions.act_window</field>
+            <field name="res_model">account.cost.center</field>
+            <field name="view_type">form</field>
+            <field name="view_mode">tree,form</field>
+            <field name="context">{'search_default_actual' : 1}</field>
```
Comment (hbrunn) [inline account_cost_center/views/account_cost_center.xml]: what's this default good for?

**Review by hbrunn — CHANGES_REQUESTED** [review]
some of the images violate [the branding rules](https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md#modules) of the OCA, in case you don't want to regenerate them, better remove them.

For all the `cost_center_id` fields in this module, I miss an index property. Is this a conscious choice?

File: account_cost_center/views/account_cost_center.xml (hunk 9 above)
Comment (astirpe) [reply account_cost_center/views/account_cost_center.xml:27]: This was meant to be used by modules that extend this one.
I was creating some new modules that add pages on this form and I thought it was useful to have a notebook here.
Let say we remove this notebook, then if you have other two modules that add pages, you will need to couple those modules: one module adds the notebook, the other module extends from the first one adding pages.

I would prefer to keep this empty elements here in order to make inheritance easier.

File: account_cost_center/models/account_invoice.py (hunk 4 above)
Comment (astirpe) [reply account_cost_center/models/account_invoice.py]: Let me try it ;)
I'll let you know!

File: account_cost_center/security/ir.model.access.csv (hunk 6 above)
Comment (astirpe) [reply account_cost_center/security/ir.model.access.csv:3]: I keep the full permissions for 'group_account_user'. Do you agree?

File: account_cost_center/views/account_cost_center.xml (hunk 8 above)
Comment (astirpe) [reply account_cost_center/views/account_cost_center.xml]: It seems this class doesn't have any effect here, I just removed the entire div.

File: account_cost_center/views/account_cost_center.xml (hunk 9 above)
Comment (hbrunn) [reply account_cost_center/views/account_cost_center.xml:27]: for a module meant as base for a lot of others, that makes sense, yes

File: account_cost_center/security/ir.model.access.csv (hunk 6 above)
Comment (hbrunn) [reply account_cost_center/security/ir.model.access.csv:3]: that's in line with having analytic accounts being fully accessible to account_user, so yes, makes sense

File: account_cost_center/models/account_move_line.py [hunk 13]
```diff
@@ -0,0 +1,15 @@
...
+
+from openerp import fields, models
+
+
+class AccountMoveLine(models.Model):
+    _inherit = 'account.move.line'
+
+    cost_center_id = fields.Many2one(
+        'account.cost.center',
+        select=True,
```
Comment (hbrunn) [inline account_cost_center/models/account_move_line.py]: it's called `index` by now: https://github.com/OCA/OCB/blob/8.0/openerp/fields.py#L139

File: account_cost_center/models/account_invoice_line.py [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+
+    @api.model
+    def _default_cost_center(self):
+        return self.env['account.cost.center'].browse(
+            self._context.get('cost_center_id'))
+
+    cost_center_id = fields.Many2one(
+        'account.cost.center',
+        string='Cost Center',
+        select=True,
```
Comment (hbrunn) [inline account_cost_center/models/account_invoice_line.py]: here too

**Comment by hbrunn** [comment]
nice! As soon as the indexes are fixed, we can merge in my opinion

**Review by hbrunn — APPROVED** [review]

**Comment by hbrunn** [comment]
nice! Any plans for further work or should I merge?

**Comment by astirpe** [comment]
For me this module is complete. Thanks!

**Comment by astirpe** [comment]
Thanks to you all for the reviews!
If you are interested, you can review the PR #442 containing the porting of account_cost_center to V9

**Pull request merged**